subsys/bluetooth/host/classic/spp.c

```diff
@@ -0,0 +1,445 @@
+		LOG_ERR("SPP rfconn connect fail, err:%d", err);
+		goto exit_err;
+	}
+
+	goto exit_stop;
+
+exit_err:
+	if (spp_cb && spp_cb->disconnected) {
+		spp_cb->disconnected(spp);
+	}
+
+exit_stop:
+	k_sem_give(&spp_sem_lock);
+	return BT_SDP_DISCOVER_UUID_STOP;
+}
+
+static struct bt_sdp_discover_params sdp_discover = {
+	.type = BT_SDP_DISCOVER_SERVICE_SEARCH_ATTR,
+	.func = bt_spp_sdp_recv,
+	.pool = &sdp_pool,
+};
+
+int bt_spp_register_cb(const struct bt_spp_cb *cb)
+{
+	if (!cb) {
+		return -EINVAL;
+	}
+
+	if (spp_cb != NULL) {
+		LOG_ERR("SPP callback already registered");
+		return -EALREADY;
+	}
+
+	spp_cb = cb;
+
+	return 0;
+}
+
+int bt_spp_register_srv(uint8_t channel)
+{
+	struct bt_spp_server *server;
+	int err;
+
+	err = spp_alloc_server(&server);
+	if (err < 0) {
+		LOG_ERR("SPP alloc server fail, err:%d", err);
+		return err;
+	}
+
+	server->rfcomm_server.channel = channel;
+	server->rfcomm_server.accept = bt_spp_accept;
+
+	bt_rfcomm_server_register(&server->rfcomm_server);
+	bt_sdp_register_service(server->sdp_rec);
+
+	return 0;
+}
+
+int bt_spp_connect(struct bt_conn *conn, struct bt_spp **spp, const struct bt_uuid *uuid)
+{
+	struct bt_spp *new_spp;
+	int err;
+
+	if (conn == NULL || spp == NULL || uuid == NULL) {
+		LOG_ERR("Invalid parameter");
+		return -EINVAL;
+	}
+
+	k_sem_take(&spp_sem_lock, K_FOREVER);
+	err = spp_alloc_connection_with_uuid(&new_spp, conn, uuid);
+	if (err < 0) {
+		k_sem_give(&spp_sem_lock);
+		LOG_ERR("SPP alloc connection fail, err:%d", err);
+		return err;
+	}
+
+	atomic_set(&new_spp->state, SPP_STATE_CONNECTING);
+
+	sdp_discover.uuid = uuid;
+	err = bt_sdp_discover(conn, &sdp_discover);
+	if (err < 0) {
+		spp_free_connection(new_spp);
+		k_sem_give(&spp_sem_lock);
+		LOG_ERR("SPP sdp discover fail");
+		return err;
+	}
+
+	*spp = new_spp;
+	return err;
+}
+
+int bt_spp_send(struct bt_spp *spp, struct net_buf *buf)
+{
+	int err;
+	uint32_t state;
+
+	if ((spp == NULL) || (buf == NULL)) {
+		LOG_ERR("spp or buf invalid");
+		return -EINVAL;
+	}
+
+	state = atomic_get(&spp->state);
+	if (state != SPP_STATE_CONNECTED) {
+		LOG_ERR("Cannot send data while not connected");
+		return -ENOTCONN;
+	}
+
+	err = bt_rfcomm_dlc_send(&spp->rfcomm_dlc, buf);
+	if (err < 0) {
+		LOG_ERR("rfcomm unable to send: %d", err);
+		net_buf_unref(buf);
+		return err;
+	}
+
+	return err;
+}
+
+int bt_spp_disconnect(struct bt_spp *spp)
+{
+	int err;
+	uint32_t state;
+
+	if (spp == NULL || spp->acl_conn == NULL) {
+		LOG_ERR("SPP or conn invalid");
+		return -EINVAL;
+	}
+
+	state = atomic_get(&spp->state);
+	if (state != SPP_STATE_CONNECTED) {
+		LOG_ERR("Cannot disconnect SPP connection while not connected");
+		return -ENOTCONN;
+	}
+
+	err = bt_rfcomm_dlc_disconnect(&spp->rfcomm_dlc);
+	if (err < 0) {
+		LOG_ERR("bt rfcomm disconnect err:%d", err);
+		return err;
+	}
+
+	atomic_set(&spp->state, SPP_STATE_DISCONNECTING);
+	return 0;
+}
+
+struct net_buf *bt_spp_create_pdu(struct bt_spp *spp, struct net_buf_pool *pool)
```

Comment:
What's the purpose of this function. It doesn't look like it's doing anything with the `spp` parameter, so in practice you don't really need any of this at all - just call `bt_rfcomm_create_pdu()` directly.